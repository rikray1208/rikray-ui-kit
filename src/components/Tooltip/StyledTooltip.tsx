import styled, { css } from 'styled-components';

import { setFontSize } from '../../utils/helpers';

import type { Placement, TooltipProps } from './index';
function setPlacement(pl: Placement) {
    switch (pl) {
        case 'Bottom':
            return css`
                --translate-y: calc(100% + var(--arrow-size));
                transform: translateX(-50%) translateY(var(--translate-y, 0)) scale(var(--scale));
                bottom: -0.25rem;
                left: 50%;
            `;
        case 'Top':
            return css`
                transform-origin: bottom center;
                --translate-y: calc(-100% - var(--arrow-size));
                transform: translateX(-50%) translateY(var(--translate-y, 0)) scale(var(--scale));
                top: -0.25rem;
                left: 50%;
            `;
        case 'Right':
            return css`
                transform-origin: left center;
                --translate-x: calc(100% + var(--arrow-size));
                transform: translateX(var(--translate-x, 0)) translateY(-50%) scale(var(--scale));
                right: -0.25rem;
                top: 50%;
            `;
        case 'Left':
            return css`
                transform-origin: right center;
                --translate-x: calc(-100% - var(--arrow-size));
                transform: translateX(var(--translate-x, 0)) translateY(-50%) scale(var(--scale));
                left: -0.25rem;
                top: 50%;
            `;
    }
}

function setTriangle(pl: Placement) {
    switch (pl) {
        case 'Top':
            return css`
                --translate-y: calc(var(--arrow-size) * -1);
                border-top-color: var(--tooltip-color);
                transform-origin: top center;
            `;
        case 'Bottom':
            return css`
                --translate-y: calc(var(--arrow-size));
                border-bottom-color: var(--tooltip-color);
                transform-origin: bottom center;
            `;
        case 'Left':
            return css`
                --translate-x: calc(var(--arrow-size) * -1);
                border-left-color: var(--tooltip-color);
                transform-origin: left center;
            `;
        case 'Right':
            return css`
                --translate-x: var(--arrow-size);
                border-right-color: var(--tooltip-color);
                transform-origin: right center;
            `;
    }
}
export const StyledTooltip = styled.div<Required<Pick<TooltipProps, 'placement'>>>`
    ${({ placement, theme }) => css`
        width: fit-content;
        height: fit-content;
        position: relative;

        &:hover:before,
        &:hover:after {
            --scale: 1;
        }

        &:before,
        &:after {
            --scale: 0;
            --tooltip-color: ${theme.Tooltip.colorBgSpotlight};
            --arrow-size: ${theme.Tooltip.arrowSize + 'px'};
            --translate-y: 0;
            --translate-x: 0;

            position: absolute;
            transform-origin: top center;
            transition: transform ${theme.Tooltip.durationFast + 'ms'} ${theme.Tooltip.transitionEaseInOut};
            ${setPlacement(placement)};
        }

        &:before {
            padding: ${theme.Tooltip['padding-1'] + 'rem'} ${theme.Tooltip['padding-2'] + 'rem'};
            width: max-content;
            text-align: center;
            background: var(--tooltip-color);
            border-radius: 8px;
            color: ${theme.Tooltip.colorWhite};
            ${setFontSize('Small', 'Tooltip', {})}

            content: attr(data-id);
        }

        &:after {
            content: '';
            border: var(--arrow-size) solid transparent;
            ${setTriangle(placement)};
        }
    `}
`;
