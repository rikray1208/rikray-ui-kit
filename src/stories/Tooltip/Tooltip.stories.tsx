import { ComponentMeta, ComponentStory } from '@storybook/react';

import React from 'react';

import { ThemeProvider, Tooltip as TooltipComp } from '../../index';

const decorators = [
    (Story: any) => (
        <ThemeProvider>
            <div>
                <Story />
            </div>
        </ThemeProvider>
    ),
];

export default {
    title: 'Design System/Global/Tooltip',
    component: TooltipComp,
    decorators: decorators,
} as ComponentMeta<typeof TooltipComp>;

const Template: ComponentStory<typeof TooltipComp> = (args) => <TooltipComp {...args} />;

export const Tooltip = Template.bind({});
Tooltip.args = {
    title: 'Some text',
    placement: 'Bottom',
    children: <p>hover me</p>,
};
