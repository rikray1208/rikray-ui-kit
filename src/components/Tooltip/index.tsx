import React, { FC } from 'react';

import { StyledTooltip } from './StyledTooltip';

export type Placement = 'Top' | 'Right' | 'Bottom' | 'Left';
export interface TooltipProps {
    title?: string;
    placement?: Placement;
    children?: React.ReactElement;
    style?: React.CSSProperties;
    className?: string;
}

const Tooltip: FC<TooltipProps> = ({ placement = 'Top', title = '', children, ...props }) => {
    return (
        <>
            <StyledTooltip {...props} placement={placement} data-id={title} data-testid='Tooltip'>
                {children}
            </StyledTooltip>
        </>
    );
};

export default Tooltip;
