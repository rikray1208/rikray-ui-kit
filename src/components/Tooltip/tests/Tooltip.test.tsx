import React from 'react';
import { create } from 'react-test-renderer';

import { ThemeProvider } from '../../../index';

import Tooltip from '../index';

describe('Tooltip', () => {
    it('should return a Tooltip', () => {
        const element = (
            <ThemeProvider>
                <Tooltip title={'Some text'} />
            </ThemeProvider>
        );
        expect(create(element).toJSON()).toMatchSnapshot();
    });
});
