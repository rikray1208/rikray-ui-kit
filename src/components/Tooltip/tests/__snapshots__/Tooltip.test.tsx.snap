// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip should return a Tooltip 1`] = `
<div
  className="sc-iTFTee ivzIZU"
  data-id="Some text"
  data-testid="Tooltip"
/>
`;
